### PR TITLE
feat: Subscribe on weights from darkvoice for smart balancing

### PR DESCRIPTION
**Pull request by kletkavrubashku**

### CODE REVIEW

File: vicodyn/src/gateway/vicodyn.cpp [hunk 1]
```diff
@@ -201,7 +203,8 @@ auto vicodyn_t::consume(const std::string& uuid,
     wrapped_gateway->consume(uuid, name, version, endpoints, protocol, extra);
     mapping.apply([&](proxy_map_t& mapping){
         if(name == "node") {
-            peers.register_peer(uuid, endpoints, extra);
+            auto peer = peers.register_peer(uuid, endpoints, extra);
+            darkmetrics.subscribe(peer->x_cocaine_cluster());
```
Comment (karitra) [inline vicodyn/src/gateway/vicodyn.cpp]: Is it any way to unsubscribe listener from cluster? Is it needed?

File: vicodyn/src/vicodyn/darkmetrics.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#include <cocaine/format/exception.hpp>
+#include <cocaine/logging.hpp>
+#include <cocaine/unicorn/value.hpp>
+
+#include <blackhole/logger.hpp>
+
+namespace cocaine {
+namespace vicodyn {
+
+darkmetrics_t::darkmetrics_t(context_t& context, peers_t& peers, const dynamic_t& args)
```
Comment (karitra) [inline vicodyn/src/vicodyn/darkmetrics.cpp]: We have agreed to have a way to disable _system_weight_ mechanics within config? Is _disable switch_ implemented already or it would be in separate PR?

File: vicodyn/src/vicodyn/darkmetrics.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        }
+        auto value = versioned_value.value();
+        auto timestamp = value.as_object()["timestamp"].as_uint();
+        for (const auto& peer : value.as_object()["nodes"].as_object()) {
+            auto system_usable_weight = peer.second.as_object()["system_usable_weight"].as_double();
+            peers_.set_system_usage_weight(peer.first, system_usable_weight);
+            COCAINE_LOG_DEBUG(logger_, "metrics for peer {}: system_usable_weight={} timestamp={}",
+                    peer.first, system_usable_weight, timestamp);
+        }
+    } catch (const std::exception& err) {
```
Comment (karitra) [inline vicodyn/src/vicodyn/darkmetrics.cpp]: Is resubscribe needed in case for unicorn/RT flap?

Comment (karitra) [inline vicodyn/src/vicodyn/darkmetrics.cpp]: In case of error old value of _system_usable_weight_ for peer would be used in peer selection.

File: vicodyn/src/vicodyn/darkmetrics.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        auto versioned_value = future.get();
+        if (versioned_value.version() < 0) {
+            COCAINE_LOG_DEBUG(logger_, "metrics by path \"{}\" is empty", path);
+            return;
+        }
+        auto value = versioned_value.value();
+        auto timestamp = value.as_object()["timestamp"].as_uint();
+        for (const auto& peer : value.as_object()["nodes"].as_object()) {
+            auto system_usable_weight = peer.second.as_object()["system_usable_weight"].as_double();
+            peers_.set_system_usage_weight(peer.first, system_usable_weight);
```
[the remark this answers is not in the document]
Comment (kletkavrubashku) [reply vicodyn/src/vicodyn/darkmetrics.cpp]: Thank you. FIxed.

File: vicodyn/src/vicodyn/darkmetrics.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            COCAINE_LOG_DEBUG(logger_, "metrics for peer {}: system_usable_weight={} timestamp={}",
```
[the remark this answers is not in the document]
Comment (kletkavrubashku) [reply vicodyn/src/vicodyn/darkmetrics.cpp]: Thank you. FIxed.

File: vicodyn/src/vicodyn/darkmetrics.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    ***/
+    try {
+        auto versioned_value = future.get();
+        if (versioned_value.version() < 0) {
+            COCAINE_LOG_DEBUG(logger_, "metrics by path \"{}\" is empty", path);
+            return;
+        }
+        auto value = versioned_value.value();
+        auto timestamp = value.as_object()["timestamp"].as_uint();
```
[the remark this answers is not in the document]
Comment (kletkavrubashku) [reply vicodyn/src/vicodyn/darkmetrics.cpp]: Thank you. FIxed.

File: vicodyn/src/gateway/vicodyn.cpp (hunk 1 above)
Comment (kletkavrubashku) [reply vicodyn/src/gateway/vicodyn.cpp]: It's very expensive to detect and seems like it is not needed.

File: vicodyn/src/vicodyn/darkmetrics.cpp (hunk 2 above)
Comment (kletkavrubashku) [reply vicodyn/src/vicodyn/darkmetrics.cpp]: Thank you, I'll add properties in config in current PR.

File: vicodyn/src/vicodyn/darkmetrics.cpp (hunk 3 above)
Comment (kletkavrubashku) [reply vicodyn/src/vicodyn/darkmetrics.cpp]: Thank you. FIxed. Old values will be changed to default after timeout.

File: vicodyn/src/vicodyn/darkmetrics.cpp (hunk 2 above)
Comment (kletkavrubashku) [reply vicodyn/src/vicodyn/darkmetrics.cpp]: Added to separate commit.

File: vicodyn/include/cocaine/gateway/vicodyn.hpp [hunk 7]
```diff
@@ -76,6 +77,7 @@ class vicodyn_t : public api::gateway_t{
     dynamic_t::object_t locator_extra;
     api::gateway_ptr wrapped_gateway;
     vicodyn::peers_t peers;
+    std::shared_ptr<vicodyn::darkmetrics_t> darkmetrics;
```
Comment (shaitan) [inline vicodyn/include/cocaine/gateway/vicodyn.hpp]: Why is `darkmetrics` stored in `std::shared_ptr`?

Comment (kletkavrubashku) [reply vicodyn/include/cocaine/gateway/vicodyn.hpp]: Because I use enable_shared_for_this to prolong life of darkmetrics.

File: vicodyn/include/cocaine/vicodyn/peer.hpp [hunk 8]
```diff
@@ -80,6 +80,8 @@ class peer_t : public std::enable_shared_from_this<peer_t> {
 
 };
 
+const double default_system_weight = 1;
```
[the remark this answers is not in the document]
Comment (kletkavrubashku) [reply vicodyn/include/cocaine/vicodyn/peer.hpp]: Sometimes I see default values in special namespace `defaults {}` and in snake_case as usual. Are you sure?

File: vicodyn/src/vicodyn/darkmetrics.cpp [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        if (ec || !lock_parent) {
+            return;
+        }
+
+        subscribe();
+        COCAINE_LOG_INFO(parent_.logger_, "resubscribed to metrics by path \"{}\"", path_);
+    });
+}
+
+darkmetrics_t::darkmetrics_t(context_t& context, peers_t& peers, const dynamic_t& args)
```
[the remark this answers is not in the document]
Comment (shaitan) [reply vicodyn/src/vicodyn/darkmetrics.cpp]: In a comment)

File: vicodyn/include/cocaine/gateway/vicodyn.hpp (hunk 7 above)
Comment (shaitan) [reply vicodyn/include/cocaine/gateway/vicodyn.hpp]: Why is it needed?

File: vicodyn/src/vicodyn/darkmetrics.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (kletkavrubashku) [reply vicodyn/src/vicodyn/darkmetrics.cpp]:
```json
{
  "unicorn": "dark",
  "unicorn_prefix": "/darkvoice",
  "unicorn_retry_after_s": 20,
  "system_weights_enabled": true,
  "system_weights_ttl_s": 300
}
```

File: vicodyn/src/vicodyn/darkmetrics.cpp [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    , unicorn_prefix_(args.as_object().at("unicorn_prefix", "/darkvoice").as_string())
+    , unicorn_retry_after_(args.as_object().at("unicorn_retry_after_s", 20U).as_uint())
+    , system_weights_enabled_(args.as_object().at("system_weights_enabled", true).as_bool())
+    , system_weights_ttl_(args.as_object().at("system_weights_ttl_s", 300U).as_uint())
+    , logger_(context.log("vicodyn/darkmetrics"))
+    , peers_(peers) {
+    COCAINE_LOG_INFO(logger_, "balancing by system weights is {}", system_weights_enabled_ ? "on" : "off");
+}
+
+darkmetrics_t::~darkmetrics_t() = default;
```
Comment (karitra) [inline vicodyn/src/vicodyn/darkmetrics.cpp]: Why you don't define it in header?

Comment (kletkavrubashku) [reply vicodyn/src/vicodyn/darkmetrics.cpp]: It will not be compiling, because default destructor of `darkmetrics_t` require destructor of `std::list<subscribtion_t>`. But I used forward declaration for `subscribtion_t`.

**Comment by kletkavrubashku** [comment]
I done `TODO: Do not prolong lifecycle of darkmetrics_t`
